### ISSUE TITLE
h2 and h2-ssl backend-protocol not working

### ISSUE THREAD

**Issue by evreng** [jcmoraisjr/haproxy-ingress#519]
Hi jcmoraisjr, everyone else involved,
thank you all for your great work on this project!
 
**IMAGE**: quay.io/jcmoraisjr/haproxy-ingress:v0.10-snapshot.3
**HAPROXY**: HA-Proxy version 2.0.12 2019/12/21

**Global confmap**
  use-htx: "true"

**Ingress annotations**
    ingress.kubernetes.io/ssl-redirect: "true"
    ingress.kubernetes.io/use-htx: "true"
    ingress.kubernetes.io/backend-protocol: h2-ssl

**_Symptoms_**: 
both the L4 and L7 healthchecks show UP, however I get a 503 "No server is available to handle this request" when I hit the frontend. 

Looking at the https backend logs:
```
2020/02/11 20:44:38 http: TLS handshake error from 192.168.17.12:57750: read tcp 10.20.76.144:8000->192.168.17.12:57750: read: connection reset by peer
```
If I change the protocol to "h1-ssl" everything works fine, proving that https itself is configured correctly. 

Ideally I would like to use cleartext "h2" also known as h2c, but since sometimes it's not well supported as HTTP/2 over TLS, I thought I'd try h2-ssl, but with no difference. 

This is a proof that the backend supports HTTP/2
```
bash-5.0# curl -vv -k https://localhost:8000
*   Trying 127.0.0.1:8000...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8000 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: C=AU; ST=Some-State; O=Internet Widgits Pty Ltd
*  start date: Feb 11 18:02:08 2020 GMT
*  expire date: Feb  8 18:02:08 2030 GMT
*  issuer: C=AU; ST=Some-State; O=Internet Widgits Pty Ltd
*  SSL certificate verify result: self signed certificate (18), continuing anyway.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x56439d151200)
> GET / HTTP/2
> Host: localhost:8000
> user-agent: curl/7.68.0
> accept: */*
> 
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* Connection state changed (MAX_CONCURRENT_STREAMS == 250)!
< HTTP/2 200 
< accept-ranges: bytes
< content-type: text/html; charset=utf-8
< last-modified: Tue, 11 Feb 2020 20:30:28 GMT
< content-length: 997
< date: Tue, 11 Feb 2020 21:16:03 GMT
< 
```

Any help would be very much appreciated.

**Comment by jcmoraisjr** [comment]
Hi, thank you for the kind words and sorry about the long delay; vacation here. Please overwrite the template with the following patch and check if it fixes this behaviour:

```diff
--- a/rootfs/etc/haproxy/template/haproxy.tmpl
+++ b/rootfs/etc/haproxy/template/haproxy.tmpl
@@ -533,7 +533,7 @@ backend {{ $backend.ID }}
 {{- define "backend" }}
     {{- $backend := .p1 }}
     {{- $server := $backend.Server }}
-    {{- if eq $server.Protocol "h2" }} proto h2{{ end }}
+    {{- if eq $server.Protocol "h2" }} proto h2 alpn h2{{ end }}
     {{- if $server.MaxConn }} maxconn {{ $server.MaxConn }}{{ end }}
     {{- if $server.MaxQueue }} maxqueue {{ $server.MaxQueue }}{{ end }}
     {{- if $server.Secure }} ssl
```

**Comment by evreng** [comment]
you rock!!
yes, this patch **does fix h2-ssl**, but h2 (h2c) still not good, same symptoms healthcheck OK but frontend returning a 503, TBH I am not sure if haproxy itself supports it at all. . .

However, this is still an excellent progress for me, since although the backend connection is encrypted, I expect a minimal SSL overhead thanks to HTTP/2 pipelining.

**Comment by evreng** [comment]
Correct me if I am wrong, according to [haproxy docs](https://cbonte.github.io/haproxy-dconv/2.0/configuration.html#5.1-proto)
```
it is possible to force the http/2 on clear TCP by specifying "proto
h2" on the bind line
```
I am just not sure where about in the template this should be amended.

**Comment by evreng** [comment]
I've read through the haproxy docs and there's no sign of h2c support, I don't think browsers support it either as yet, I've got a backend that supports h2c and was hoping to use it but nevermind. 
@jcmoraisjr I think you may want to remove it as a backend-protocol option?

**Comment by jcmoraisjr** [comment]
Hi thanks for the feedback! This is still a bug and I'm reopening this issue. =) Btw thanks also for reporting it.

I do use haproxy ingress v0.10 to ssl-offload a grpc workload without ssl, so I can confirm `h2` does work like a charm. `h2-ssl` however is broken due to the missing alpn config - the tls handshake said we want to talk http/1.1 and the L7 header is http/2.

**Comment by san360** [comment]
Whats the plan to release this? We need to leverage gRPC at our ingress. Could you please share some timelines?

**Comment by jcmoraisjr** [comment]
Hi, I forgot to close this one. This was fixed in `v0.9-beta.3` which we consider production quality. `v0.9` should be released within a few days. Closing.
